### DISCUSSION TITLE
Formulas do not get updated depending on position of cells

### DISCUSSION THREAD

**Discussion by arkni8** [exceljs/exceljs#2128] [Q&A]
I have a Excel file where I need to enter certain values from a JSON file. The excel sheet has pre-defined formulas that will use the values I will enter from JSON file using Exceljs.

Now I know that Exceljs doesn't do calculation. But here's the problem. Let's take an example to understand better.

A1 = 5
B1 = 4
C1 = 7
E1 = (A1+B1+C1) / 3       *----> which means the value of E1 = 5.33*

Excel (or any spreadsheet really) gives us the convenience of dragging a formula cell and spreading it down the column and the formula automatically changes to, say,

E2 = (A2+B2+C2) / 3
E3 = (A3+B3+C3) / 3
E4 = (A4+B4+C4) / 3
*and so on...*

But, when I access these cells in exceljs, the formula they show is *E2 = (A1+B1+C1) / 3; E3 = (A1+B1+C1) / 3*.
And their shareType is set to shared.

Is there anyway to update these shared formula depending on position of cell.

These cells only get updated to correct formula when I open the file in excel and then very intentionally click on the formula and accept the formula. Then the formula is recognized by exceljs correctly and the formula is not 'shared' anymore in the shareType.

Any feedback or help here will be appreciated. If you know of another library which can recognize these kind of shared formula will also be helpful. Thanks.

**Comment by GlennQuirynen** [comment]
Hi,

having probably similar issues with exceljs on your remark "Now I know that Exceljs doesn't do calculation".

Can you elaborate on this statement?  What do you actually mean by this?

As background, we want to achieve the following

we want to 
- dynamically (in code) set the formula for a certain cell. (eg worksheet.getCell('B1').value = { formula: 'A2'}, or how should this be set?)
- and then read the value of that cell, which in essence is outcome of the formula (eg worksheet.getCell('B1').value)

Am I now understanding correctly that this package does not support this basic behavior?
Any other packages who do?

**Reply by arkni8** [reply]
You got it Glenn. Pretty much as you summarize. But as you probably already understand, Exceljs doesn't do the calculations. So when do these formulas get calculated?
It happens when you actually open a spreadsheet (in LibreOffice or whatever you use) and that's when all the fields are calculated. Alternatively, if fields don't get updated, you can force the spreadsheet program to recalculate all the fields.

Until you open the file in a spreadsheet program, these fields only contain a formula, no result. Having said that, additionally, my problem was slightly more deeper than just this.

Say, you open a spreadsheet file, brand new, populate A1 with 5. Populate B1 with 6. And then populate C1 with formula A1+ B1. AGAIN, THIS IS IN A SPREADSHEET APP, like MS Office. Now, almost all Spreadsheet apps have a convenience feature of dragging a formula box down a column or across a row, right? It propagates formula predictably so we don't have to fill each field in a column or row with formula like an absolute chore. Now we can check formula in C2, it should be A2 + B2. For C3, the formula will be A3 + B3. I think you get it. So we are done with the spreadsheet app now.

Now we move to Exceljs in Nodejs. Try populating A2 and B2 now with value 8 and 2 respectively. What do you think will happen? Remember. Exceljs DOESN'T calculate. So, nothing happens. It just populates A2 and B2. That's it. I understand that part. But if you check the formula of C2 at this moment, what do you think it will show? You expect it to show A2 + B2, right? Well it does not. It will show a 'shared' formula of A1 + B1. And this is my problem. Perhaps its just how spreadsheets save data, by associating with the nearest populated value, but it did throw my plan under the bus.

Did I find workaround? Or another package?
Answer: Yes, I found a workaround. No, it doesn't involve another package.
Short answer: Just open your file and force save it once before trying to read the fields with formula. Yea, you will not be able to FULLY automate this part unless you have a script running somewhere to orchestrate everything.

Long Answer: As I mentioned before, Exceljs doesn't compute. Its the Spreadsheet apps that compute. So when we open the spreadsheet (after populating A2 and B2 in exceljs) in a spreadsheet app like MS Office or LibreOffice, the program recalculates all field. Force recalculating sometimes helped as well. So all fields get recalculated and you save it. You do not have to touch any fields, nothing. Just recalculate and save the file. Will take less than 5 second if the spreadsheet is of reasonable size. Now go back to Exceljs and try accessing the C2 field. 

```
console.log( worksheet.getCall('C2').value )
// Output -> { formula: 'A2 + B2' , result: '10' }

console.log( worksheet.getCall('C3').value )
// Output -> { formulaType: 'shared', formula: 'A1 + B1' }
// I actually do not remember if you get an object with property 'formulaType' in this way - but Exceljs does inform you 
whatever formula you are trying to access is shared with a 'master' cell. You can read about shared formula in 
documentation actually.
```
You can see the different between C2 and C3, right? The field that could get calculated in the spreadsheet app now shows the correct formula, and is not a 'shared' formula anymore whereas, C3 (and all the formula fields that couldn't be calculated) stay with their shared formula.

Hope this gives you more insight. I know I answered it a bit late haha. 

But if someone in the community has some answers to this behavior, let us know.

**Comment by gregkonush** [comment]
Update for anyone landing here from ExcelJS search: ExcelJS is still the file I/O layer, not the formula engine. If the backend path is "change inputs, recalc formulas, read proof values" without opening Excel or LibreOffice, the scoped bridge is here:

https://www.npmjs.com/package/@bilig/exceljs-formula-recalc

It keeps ExcelJS as the workbook boundary, runs Bilig's XLSX import/recalc/export path, reloads the recalculated bytes into the workbook, and patches requested readback cells.

```ts
import ExcelJS from "exceljs";
import { recalculateExceljsWorkbook } from "@bilig/exceljs-formula-recalc";

const workbook = new ExcelJS.Workbook();
await workbook.xlsx.readFile("input.xlsx");

const result = await recalculateExceljsWorkbook(workbook, {
  edits: [{ target: "Sheet1!A1", value: 9 }],
  reads: ["Sheet1!C1"],
});

console.log(result.reads["Sheet1!C1"]);
console.log(workbook.getWorksheet("Sheet1")?.getCell("C1").value);
```

If you are not already using ExcelJS objects, the lower-friction package is `@bilig/xlsx-formula-recalc`. If the problem is really backend spreadsheet-shaped business logic rather than XLSX file I/O, start with `@bilig/workpaper`.

I would still validate shared-formula-heavy or Excel-compatibility-sensitive workbooks against Excel/LibreOffice before treating it as a drop-in replacement for desktop Excel behavior. This is mainly for Node services that need deterministic readback after mutating inputs.

Disclosure: I maintain Bilig.
